### ISSUE TITLE
Provide the ability to get the notification object in PublishCore

### ISSUE THREAD

**Issue by ahocquet** [jbogard/MediatR#429]
I've developed a `Polly` based custom Mediator implementation, in order to add exponential retries and fallback. Therefore, it never throws an exception to the `Publish` caller.

The overriden PublishCore is clean and simple : 
```
protected override async Task PublishCore(IEnumerable<Func<Task>> allHandlers)
        {
            var tasks = allHandlers.Select(t => _policy.ExecuteAsync(t));
            await Task.WhenAll(tasks).ConfigureAwait(false);
        }
```

The Polly policy is defined in the constructor, but it could also be injected. 

When my `Fallback` policy triggers, I would like to send the failed notification to an Azure Service Bus Queue, so I can monitor it and replay it later through an Azure Function for example.

In order to achieve that, I would need to get access to the underlying `INotification` object in the `PublishCore` method. I see two possibilities:
- Inject a `IEnumerable<Expression<Func<Task>>>` instead of a `IEnumerable<Func<Task>>`. Thanks to reflection, I would get access to the `INotification` object.
- Introduce a breaking change by updating the `PublishCore` signature in order to provide an `INotification notification` parameter, even though it wouldn't be used in the default implementation.

If you're ok with the idea, I can send a PR based on the preferred solution.

**Comment by ahocquet** [comment]
I could get it done by implementing this logic in a handler decorator instead of a custom Mediator implementation. It works great, so I don't know if my request is still relevant. 

Maybe it's better to keep this logic apart, and maintaining the single responsibility principle. But on the other side, it comes with some plumbing that not all containers support (Asp.net core container, for example, doesn't have decorators support out of the box).

**Comment by bartosz6** [comment]
#353 linked issue

I think a decorator is clean a good solution. No need to modify the mediatr.

**Comment by ahocquet** [comment]
Very well then, let's keep it that way.

**Comment by jbogard** [comment]
FYI this next release will have a breaking change anyway, so if we want to introduce a breaking change, now would be the time.

@bartosz6 I can see decorators being a possibility - but one of the reasons I added the whole behavior pipeline was because of sparse support in DI containers for decorators.

**Comment by ahocquet** [comment]
@jbogard If you think this feature makes sense, I'll be pleased to send you a PR. :smile:

**Comment by jbogard** [comment]
I think it's reasonable - people have asked for behaviors for notifications but to me that never made much sense, because some folks want behaviors around individual handlers vs. ALL the handlers.

This seems like a reasonable first step until we get more information about what is needed.

**Comment by jbogard** [comment]
And I'm open to a PR on this one

**Comment by jbogard** [comment]
Fixed by #431
